File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.java
class Solution {
    public long mostPoints(int[][] questions) {
        int n = questions.length;
        if(n == 1) return questions[0][0];
        // dp[i] represents the maximum points that can be get from ith question to nth question. 
        long[] dp = new long[n];
        // move from index n to index 0;
        dp[n-1] = questions[n-1][0];
        for(int i = n-2; i >= 0; i--){
            long rest = i + questions[i][1] + 1 < n ? dp[i + questions[i][1] + 1] : 0;
            dp[i] = Math.max(dp[i+1], questions[i][0] + rest);
        }
        return dp[0];
    }
}

File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.cpp
class Solution {
public:
    long long dfs(int ind, vector<vector<int>>& questions, int n, vector<long long>& dp) {
        if (ind >= n) return 0;
        if (dp[ind] != -1) return dp[ind];
        long long pick = questions[ind][0] + dfs(ind + questions[ind][1] + 1, questions, n, dp);
        long long notPick = 0 + dfs(ind + 1, questions, n, dp);

        return dp[ind] = max(pick, notPick);
    }
    long long mostPoints(vector<vector<int>>& questions) {
        int n = questions.size();
        vector<long long> dp(n, -1);
        return dfs(0, questions, n, dp);
    }
};